### ISSUE TITLE
`MacawFreshSymbolic: XXX type (bv 64)` crash on AArch32 program using `strd` instruction

### ISSUE THREAD

**Issue by RyanGlScott** [GaloisInc/macaw#301]
Given the following assembly code, which was extracted from uClibc:

```
@ test.s

  .global my_seed48
  .syntax unified
my_seed48:
  ldr     r3, [pc, #28]
  ldr     r2, [pc, #28]
  push    {r4, lr}
  add     r3, pc, r3
  ldr     r4, [r3, r2]
  mov     r1, r4
  bl      my__GI_seed48_r
  add     r0, r4, #6
  pop     {r4, pc}
.4byte 0x000182f8
.4byte 0x000009cc

  .global my__GI_seed48_r
  .syntax unified
my__GI_seed48_r:
  push    {r3, r4, r5, lr}
  mov     r2, #6
  mov     r5, r0
  add     r0, r1, #6
  mov     r4, r1
  bl      memcpy
  ldrh    r3, [r5, #4]
  mov     r0, #0
  ldr     r2, [pc, #44]
  strh    r3, [r4, #4]
  ldrh    r3, [r5, #2]
  strh    r3, [r4, #2]
  mov     r3, #5
  ldrh    r5, [r5]
  strd    r2, [r4, #16]
  mov     r3, #11
  strh    r3, [r4, #12]
  mov     r3, #1
  strh    r5, [r4]
  strh    r3, [r4, #14]
  pop     {r3, r4, r5, pc}
.4byte 0xdeece66d
```

And the following test harness:

```c
// test.c
extern unsigned short *my_seed48(unsigned short seed16v[3]);

int __attribute__((noinline)) test_seed48() {
  unsigned short seed16v[3] = {0,0,0};
  my_seed48(seed16v);
  return 0;
}

void _start() {
  test_seed48();
}
```

Compile this into an executable using an ARM `musl-gcc` like so:

```
$ armv7m-linux-musleabi-gcc -nostartfiles -static -fno-stack-protector -no-pie -mcpu=cortex-a7 test.s -c -o test.o
$ armv7m-linux-musleabi-gcc -nostartfiles -static -fno-stack-protector -no-pie -mcpu=cortex-a7 test.c test.o -o test.exe
```

And run it with the following Haskell program:

<details>

```hs
-- Bug.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE ImplicitParams #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeOperators #-}
module Main (main) where

import           Control.Lens ( (^.) )
import qualified Data.ByteString as BS
import qualified Data.ByteString.Char8 as BS8
import qualified Data.ElfEdit as Elf
import qualified Data.Foldable as F
import qualified Data.Map as Map
import           Data.Maybe ( mapMaybe )
import qualified Data.Parameterized.Classes as PC
import qualified Data.Parameterized.Nonce as PN
import           Data.Parameterized.Some ( Some(..) )
import           Data.Proxy ( Proxy(..) )

import qualified Language.ASL.Globals as ASL
import qualified Data.Macaw.Architecture.Info as MAI

import           Data.Macaw.AArch32.Symbolic ()
import qualified Data.Macaw.ARM as MA
import qualified Data.Macaw.ARM.ARMReg as MAR
import qualified Data.Macaw.Discovery as M
import qualified Data.Macaw.Symbolic as MS
import qualified Data.Macaw.Symbolic.Testing as MST
import qualified What4.Config as WC
import qualified What4.Expr as WE
import qualified What4.Expr.Builder as WEB
import qualified What4.Interface as WI
import qualified What4.ProblemFeatures as WPF
import qualified What4.Solver as WS

import qualified Lang.Crucible.Backend as CB
import qualified Lang.Crucible.Backend.Online as CBO
import qualified Lang.Crucible.Simulator as CS
import qualified Lang.Crucible.LLVM.MemModel as LLVM

main :: IO ()
main = do
  -- These are pass/fail in that the assertions in the "pass" set are true (and
  -- the solver returns Unsat), while the assertions in the "fail" set are false
  -- (and the solver returns Sat).
  let passRes = MST.SimulationResult MST.Unsat
  mkSymExTest passRes "test.exe"

hasTestPrefix :: Some (M.DiscoveryFunInfo arch) -> Maybe (BS8.ByteString, Some (M.DiscoveryFunInfo arch))
hasTestPrefix (Some dfi) = do
  bsname <- M.discoveredFunSymbol dfi
  if BS8.pack "test_" `BS8.isPrefixOf` bsname
    then return (bsname, Some dfi)
    else Nothing

mkSymExTest :: MST.SimulationResult -> FilePath -> IO ()
mkSymExTest expected exePath = do
  bytes <- BS.readFile exePath
  case Elf.decodeElfHeaderInfo bytes of
    Left (_, msg) -> fail ("Error parsing ELF header from file '" ++ show exePath ++ "': " ++ msg)
    Right (Elf.SomeElf ehi) -> do
      case Elf.headerClass (Elf.header ehi) of
        Elf.ELFCLASS32 ->
          symExTestSized expected ehi MA.arm_linux_info
        Elf.ELFCLASS64 -> fail "64 bit ARM is not supported"

symExTestSized :: MST.SimulationResult
               -> Elf.ElfHeaderInfo 32
               -> MAI.ArchitectureInfo MA.ARM
               -> IO ()
symExTestSized expected ehi archInfo = do
   (mem, discState) <- MST.runDiscovery ehi MST.toAddrSymMap archInfo
   let funInfos = Map.elems (discState ^. M.funInfo)
   let testEntryPoints = mapMaybe hasTestPrefix funInfos
   F.forM_ testEntryPoints $ \(name, Some dfi) -> do
     putStrLn ("Testing " ++ BS8.unpack name ++ " at " ++ show (M.discoveredFunAddr dfi))
     Some (gen :: PN.NonceGenerator IO t) <- PN.newIONonceGenerator
     sym <- WEB.newExprBuilder WEB.FloatRealRepr WE.EmptyExprBuilderState gen
     CBO.withYicesOnlineBackend sym CBO.NoUnsatFeatures WPF.noFeatures $ \bak -> do
       -- We are using the z3 backend to discharge proof obligations, so
       -- we need to add its options to the backend configuration
       let solver = WS.z3Adapter
       let backendConf = WI.getConfiguration sym
       WC.extendConfig (WS.solver_adapter_config_options solver) backendConf

       execFeatures <- MST.defaultExecFeatures (MST.SomeOnlineBackend bak)
       let Just archVals = MS.archVals (Proxy @MA.ARM) Nothing
       let extract = armResultExtractor archVals
       let ?memOpts = LLVM.defaultMemOptions
       simRes <- MST.simulateAndVerify solver WS.defaultLogData bak execFeatures archInfo archVals mem extract discState dfi
       if expected == simRes
         then putStrLn "Test passed"
         else fail $ "Expected " ++ show expected ++ ", but got " ++ show simRes

-- | ARM functions with a single scalar return value return it in %r0
--
-- Since all test functions must return a value to assert as true, this is
-- straightforward to extract
armResultExtractor :: ( CB.IsSymInterface sym
                      )
                   => MS.ArchVals MA.ARM
                   -> MST.ResultExtractor sym MA.ARM
armResultExtractor archVals = MST.ResultExtractor $ \regs _sp _mem k -> do
  let re = MS.lookupReg archVals regs (MAR.ARMGlobalBV (ASL.knownGlobalRef @"_R0"))
  k PC.knownRepr (CS.regValue re)
```
</details>

Then it will produce the following crash:

```
$ runghc-9.2 Bug.hs
Testing test_seed48 at 0x10074
Bug.hs: MacawFreshSymbolic: XXX type (bv 64)
CallStack (from HasCallStack):
  error, called at src/Data/Macaw/Symbolic.hs:1156:21 in macaw-symbolic-0.0.1-inplace:Data.Macaw.Symbolic
```

**Comment by travitch** [comment]
Those are indirectly generated by the `freshSymbolic` combinator: https://github.com/GaloisInc/macaw/blob/8bce1cf7693afd6a5ef91a8acb8d6dfe5fd70b45/symbolic/src/Data/Macaw/Symbolic/CrucGen.hs#L1285

That combinator is called from the translation of `SetUndefined`: https://github.com/GaloisInc/macaw/blob/8bce1cf7693afd6a5ef91a8acb8d6dfe5fd70b45/symbolic/src/Data/Macaw/Symbolic/CrucGen.hs#L1300

Those are generated in the AArch32 semantics: https://github.com/GaloisInc/macaw/blob/8bce1cf7693afd6a5ef91a8acb8d6dfe5fd70b45/macaw-aarch32/src/Data/Macaw/ARM/Semantics/TH.hs#L411-L414

**Comment by travitch** [comment]
Looking at the instruction semantics, I see (excerpted):

``` 
("STRD_i_A1_off"
   ((SigEnv
    (("ufn9"
     ("df_getSlice_4N_4length_1_norm"
      ((BV 4)
       Bool
       (BV 65))
      (Struct
       ((BV 1)))))
     ("ufn21"
      ("df_ConditionHolds_1_norm"
       ((BV 4)
        (BV 1)
        (BV 1)
        (BV 1)
        (BV 1))
       (Struct
        (Bool))))
     ("ufn32"
      ("df_BAREGETTER_PC_0_norm"
       ((BV 1)
        (BV 1)
        (BV 32))
       (Struct
        ((BV 32)))))
     ("ufn38"
      ("uf_gpr_get"
       ((BV 148)
        (BV 4))
       (BV 32)))
     ("ufn54"
      ("uf_INIT_GLOBAL_HCR_EL2"
       ()
       (BV 64)))
     ("ufn58"
      ("uf_INIT_GLOBAL_SCR"
       ()
       (BV 32)))
     ("ufn60"
      ("uf_INIT_GLOBAL_SCR_EL3"
       ()
       (BV 32)))
     ("ufn62"
      ("uf_INIT_GLOBAL_SCTLR_EL1"
       ()
       (BV 64)))
     ("ufn64"
      ("uf_INIT_GLOBAL_SCTLR_EL2"
       ()
       (BV 64)))
     ("ufn66"
      ("uf_INIT_GLOBAL_SCTLR_EL3"
       ()
       (BV 64)))

```
Those `uf_INIT_GLOBAL` cases are the ones seen above - if those are referenced (and they probably are if they show up like that), the undefined values could be coming from them. We may need to fix the values of some of these to something inoffensive to avoid referencing them

**Comment by RyanGlScott** [comment]
That is an interesting possibility. Then again, I scoured the functions that `STRD_i_A1_off` transitively invokes, and for the life of me, I can't find any references to any of the global variables mentioned above.

But after taking another look at the ASL source code for `STRD_i_A1_off`, I have a much more likely explanation for what is going on:

```
    __execute __conditional
        offset_addr = if add then (R[n] + imm32) else (R[n] - imm32);
        address = if index then offset_addr else R[n];
        if address == Align(address, 8) then
            bits(64) data;
            if BigEndian() then
                data<63:32> = R[t];
                data<31:0> = R[t2];
            else
                data<31:0> = R[t];
                data<63:32> = R[t2];
            MemA[address,8] = data;
        else
            MemA[address,4] = R[t];
            MemA[address+4,4] = R[t2];
        if wback then R[n] = offset_addr;
```

I'm 99% sure that the culprit is the `bits(64) data;` statement. `asl-translator` will translate declared-but-not-defined identifiers like `data` such that they are initialized with `uf_UNDEFINED_<type>` (see [this code](https://github.com/GaloisInc/asl-translator/blob/4a7eb74dada86f5296654e0daeecc21ae4e94504/lib/Language/ASL/Translation.hs#L318)). Semantics-wise, this seems plausible, since the following lines use slices to initialize all of `data`'s bits. The implementation of the semantics in `macaw` falls over in practice, however, since `uf_UNDEFINED_<type>` is converted to `MacawFreshSymbolic`, and our semantics for `MacawFreshSymbolic` will simply throw an exception if `<type>` is anything but a `boolean`.

**Comment by travitch** [comment]
Interesting. So we could try to just implement `MacawFreshSymbolic` and hope that the symbolic terms fall out. That seems like the easy first thing to try. I suppose we don't have a better initializer.

**Comment by travitch** [comment]
Yeah my worry is that we might get something like padding left undefined, and have that show up in queries later. On one hand that is accurate, on the other it really sucks.

I'd say lets try with just a raw bitvector and see what happens... we'll know immediately if it is the wrong choice.

**Comment by RyanGlScott** [comment]
Here is a skeleton of a patch:

```diff
diff --git a/symbolic/src/Data/Macaw/Symbolic.hs b/symbolic/src/Data/Macaw/Symbolic.hs
index c980aa93..488a00b5 100644
--- a/symbolic/src/Data/Macaw/Symbolic.hs
+++ b/symbolic/src/Data/Macaw/Symbolic.hs
@@ -155,9 +155,11 @@ import           Data.Parameterized.Some ( Some(Some) )
 import qualified Data.Parameterized.TraversableFC as FC
 import qualified Data.Set as S
 import qualified Data.Text as T
+import qualified Data.Vector as V
 
 import qualified What4.FunctionName as C
 import           What4.Interface
+import           What4.InterpretedFloatingPoint (freshFloatConstant)
 import qualified What4.Utils.StringLiteral as C
 import qualified What4.ProgramLoc as WPL
 
@@ -1147,13 +1149,8 @@ execMacawStmtExtension (SB.MacawArchEvalFn archStmtFn) mvar globs (MO.LookupFunc
     MacawGlobalPtr w addr ->
       M.addrWidthClass w $ doGetGlobal st mvar globs addr
 
-    MacawFreshSymbolic t -> -- XXX: user freshValue
-      do nm <- case userSymbol "macawFresh" of
-                 Right a -> return a
-                 Left err -> fail (show err)
-         v <- case t of
-               M.BoolTypeRepr -> freshConstant sym nm C.BaseBoolRepr
-               _ -> error ("MacawFreshSymbolic: XXX type " ++ show t)
+    MacawFreshSymbolic t ->
+      do v <- freshCrucibleConstant sym (safeSymbol "macawFresh") t
          return (v,st)
 
     MacawLookupFunctionHandle _ args -> do
@@ -1181,6 +1178,32 @@ execMacawStmtExtension (SB.MacawArchEvalFn archStmtFn) mvar globs (MO.LookupFunc
     PtrTrunc w x                -> doPtrTrunc st mvar x w
     PtrUExt w x                 -> doPtrUExt st mvar x w
 
+freshCrucibleConstant ::
+  forall sym tp.
+  IsSymInterface sym =>
+  sym ->
+  SolverSymbol ->
+  M.TypeRepr tp ->
+  IO (C.RegValue sym (ToCrucibleType tp))
+freshCrucibleConstant sym nm = go
+  where
+    go :: forall tp'.
+          M.TypeRepr tp' ->
+          IO (C.RegValue sym (ToCrucibleType tp'))
+    go M.BoolTypeRepr =
+      freshConstant sym nm C.BaseBoolRepr
+    go (M.BVTypeRepr w) = do
+      off <- freshConstant sym nm (C.BaseBVRepr w)
+      MM.llvmPointer_bv sym off
+      -- blk <- freshNat sym nm
+      -- off <- freshConstant sym nm (C.BaseBVRepr w)
+      -- pure $ MM.LLVMPointer blk off
+    go (M.FloatTypeRepr fi) =
+      freshFloatConstant sym nm (floatInfoToCrucible fi)
+    go (M.TupleTypeRepr l) = macawListToCrucibleM (fmap C.RV . go) l
+    go (M.VecTypeRepr n tp) =
+      V.replicateM (fromIntegral (natValue n)) (go tp)
 
 -- | Return macaw extension evaluation functions.
 macawExtensions
```

This implements the raw-bitvector-with-symbolic offset approach, but I have also included the fully-symbolic-pointer approach as comments. In the process of doing this, I discovered that the test case that I wrote about isn't quite self-contained, since the implementation of `seed48` references an `R_ARM_GLOB_DAT` relocation that I didn't include. In lieu of trying to extract this relocation, I simply tested the above patch on uClibc's `seed48` function directly using a `macaw-symbolic`–based verifier. Happily, both approaches appear to work equally well, as in both cases the crash disappears and the verifier is able to prove all of the goals that arise from simulating `seed48`.

**Comment by RyanGlScott** [comment]
Ah, here is a simpler test case:

```c
#include <stdint.h>

uint64_t x = 0;

int __attribute__((noinline)) test_strd() {
  x = 42;
  return x == 42;
}

void _start() {
  test_strd();
}
```
```
$ arm-linux-gnueabi-gcc -nostdlib -static -fno-stack-protector -mcpu=cortex-a7 -O0 test-strd.c -o test-strd.unopt.exe
$ $ arm-linux-gnueabi-objdump -d test-strd.unopt.exe 

test-strd.unopt.exe:     file format elf32-littlearm


Disassembly of section .text:

000100d8 <test_strd>:
   100d8:       e52db004        push    {fp}            ; (str fp, [sp, #-4]!)
   100dc:       e28db000        add     fp, sp, #0
   100e0:       e3003138        movw    r3, #312        ; 0x138
   100e4:       e3403002        movt    r3, #2
   100e8:       e3a0002a        mov     r0, #42 ; 0x2a
   100ec:       e3a01000        mov     r1, #0
   100f0:       e1c300f0        strd    r0, [r3]
   100f4:       e3003138        movw    r3, #312        ; 0x138
   100f8:       e3403002        movt    r3, #2
   100fc:       e1c320d0        ldrd    r2, [r3]
   10100:       e3530000        cmp     r3, #0
   10104:       0352002a        cmpeq   r2, #42 ; 0x2a
   10108:       03a03001        moveq   r3, #1
   1010c:       13a03000        movne   r3, #0
   10110:       e6ef3073        uxtb    r3, r3
   10114:       e1a00003        mov     r0, r3
   10118:       e28bd000        add     sp, fp, #0
   1011c:       e49db004        pop     {fp}            ; (ldr fp, [sp], #4)
   10120:       e12fff1e        bx      lr

00010124 <_start>:
   10124:       e92d4800        push    {fp, lr}
   10128:       e28db004        add     fp, sp, #4
   1012c:       ebffffe9        bl      100d8 <test_strd>
   10130:       e320f000        nop     {0}
   10134:       e8bd8800        pop     {fp, pc}
```

That emits a `strd` instruction without any relocation nonsense. Unfortunately, this version _still_ doesn't pass when placed into the `macaw-aarch32-symbolic` test suite when implementing either approach from https://github.com/GaloisInc/macaw/issues/301#issuecomment-1182242388:

```
    tests/pass/test-strd.unopt.exe:               FAIL (0.35s)
      Testing test_strd at 0x100d8 (0.08s)
        tests/Main.hs:153:
        AssertionResult
        expected: SimulationResult Unsat
         but got: SimulationResult Sat
      
      Use -p '/tests\/pass\/test-strd.unopt.exe/' to rerun this test only.
```

Interestingly, it _does_ pass if you compile it with optimizations:

```
$ arm-linux-gnueabi-gcc -nostdlib -static -fno-stack-protector -mcpu=cortex-a7 -O2 test-strd.c -o test-strd.opt.exe
$ arm-linux-gnueabi-objdump -d test-strd.opt.exe

test-strd.opt.exe:     file format elf32-littlearm


Disassembly of section .text:

000100d8 <test_strd>:
   100d8:       e3003100        movw    r3, #256        ; 0x100
   100dc:       e3a00001        mov     r0, #1
   100e0:       e3403002        movt    r3, #2
   100e4:       e92d0030        push    {r4, r5}
   100e8:       e3a0402a        mov     r4, #42 ; 0x2a
   100ec:       e3a05000        mov     r5, #0
   100f0:       e1c340f0        strd    r4, [r3]
   100f4:       e8bd0030        pop     {r4, r5}
   100f8:       e12fff1e        bx      lr

000100fc <_start>:
   100fc:       eafffff5        b       100d8 <test_strd>
```

**Comment by RyanGlScott** [comment]
Incredibly, a small tweak to the program above reveals yet _another_ bug. See #302.

**Comment by RyanGlScott** [comment]
Happily, the unoptimized test case in https://github.com/GaloisInc/macaw/issues/301#issuecomment-1183555233 now passes with the latest `macaw`. Since the unoptimized program uses the `movw` instruction, I suspect that this is a result of the changes in https://github.com/GaloisInc/asl-translator/pull/47—hooray!

I've submitted this fix (and the accompanying test cases) in #351.
